File: backend/src/snowan/server/mcp.py
"""MCP servers: list / add (form or paste-JSON) / enable / per-tool policy / remove
/ probe (connect + list a server's tools)."""
from fastapi import APIRouter, HTTPException
from fastapi.responses import Response
from pydantic import BaseModel

from .. import mcp as store
from ..agent.mcp import probe

router = APIRouter(prefix="/api/mcp")


class ServerForm(BaseModel):
    name: str
    command: str = ""
    args: list[str] = []
    env: dict = {}
    url: str = ""
    headers: dict = {}


class PasteBody(BaseModel):
    config: dict  # full {"mcpServers": {...}} or a single-entry object


class EnableBody(BaseModel):
    enabled: bool


class PolicyBody(BaseModel):
    tool: str
    policy: str  # ask | auto | off


@router.get("")
def list_servers() -> list[dict]:
    return store.list_servers()


@router.post("")
def add_server(req: ServerForm) -> dict:
    if not req.name.strip():
        raise HTTPException(422, "需要名称")
    if not (req.command.strip() or req.url.strip()):
        raise HTTPException(422, "stdio 需要命令,远程需要 URL")
    cfg = (
        {"command": req.command.strip(), "args": req.args, "env": req.env}
        if req.command.strip()
        else {"url": req.url.strip(), "headers": req.headers}
    )
    store.upsert_server(req.name.strip(), cfg)
    return {"ok": True}


@router.post("/paste")
def paste(req: PasteBody) -> dict:
    try:
        names = store.merge_paste(req.config)
    except ValueError as e:
        raise HTTPException(422, str(e))
    return {"ok": True, "names": names}


@router.patch("/{name}/enabled")
def set_enabled(name: str, req: EnableBody) -> dict:
    store.set_enabled(name, req.enabled)
    return {"ok": True}


@router.patch("/{name}/tool-policy")
def set_policy(name: str, req: PolicyBody) -> dict:
    try:
        store.set_tool_policy(name, req.tool, req.policy)
    except ValueError as e:
        raise HTTPException(422, str(e))
    return {"ok": True}


@router.delete("/{name}")
def remove(name: str) -> Response:
    store.remove_server(name)
    return Response(status_code=204)


@router.post("/{name}/probe")
async def probe_server(name: str) -> dict:
    return await probe(name)

File: frontend/src/api/mcp.ts
import { api } from './base';

export type ToolPolicy = 'ask' | 'auto' | 'off';

export type McpServer = {
  name: string;
  transport: 'stdio' | 'http';
  enabled: boolean;
  command: string;
  args: string[];
  env: Record<string, string>;
  url: string;
  headers: Record<string, string>;
  tools: Record<string, ToolPolicy>;
};

export type McpProbe = {
  ok: boolean;
  tools?: { name: string; description: string }[];
  error?: string;
};

const j = <T>(r: Response): Promise<T> => {
  if (!r.ok) throw new Error(`${r.status}`);
  return r.json();
};

const send = <T>(method: string, path: string, body?: unknown): Promise<T> =>
  fetch(api(path), {
    method,
    headers: body ? { 'Content-Type': 'application/json' } : undefined,
    body: body ? JSON.stringify(body) : undefined,
  }).then((r) => j<T>(r));

export const listServers = () => fetch(api('/api/mcp')).then((r) => j<McpServer[]>(r));

export const addServer = (form: {
  name: string;
  command?: string;
  args?: string[];
  env?: Record<string, string>;
  url?: string;
  headers?: Record<string, string>;
}) => send<{ ok: boolean }>('POST', '/api/mcp', form);

export const pasteConfig = (config: unknown) =>
  send<{ ok: boolean; names: string[] }>('POST', '/api/mcp/paste', { config });

export const setEnabled = (name: string, enabled: boolean) =>
  send<{ ok: boolean }>('PATCH', `/api/mcp/${encodeURIComponent(name)}/enabled`, { enabled });

export const setToolPolicy = (name: string, tool: string, policy: ToolPolicy) =>
  send<{ ok: boolean }>('PATCH', `/api/mcp/${encodeURIComponent(name)}/tool-policy`, { tool, policy });

export const removeServer = (name: string) =>
  fetch(api(`/api/mcp/${encodeURIComponent(name)}`), { method: 'DELETE' });

export const probe = (name: string) =>
  send<McpProbe>('POST', `/api/mcp/${encodeURIComponent(name)}/probe`);
